### ISSUE TITLE
Able to rename a snapshot name by renaming the directories under .zfs/shares?

### ISSUE THREAD

**Issue by mailinglists35** [openzfs/zfs#13423]
### Describe the feature would like to see added to OpenZFS

Today I entered a dot zfs directory and while looking around I decided that one of the snapshots should be prevented from being deleted.
_It's going to be deleted by good old `zfs-auto-snapshot` shell script so in order to save it from deletion a rename is needed to no longer contain the zfs-auto-snap string._ 
And it felt so naturally to type the mv command until I realized its not going to work and I would have to go to 'zfs rename' command.
But I am thinking, is it technically possible so it won't remain just my imagination?

**Comment by jittygitty** [comment]
@mailinglists35 I never used zfs-auto-snapshot, so not sure what's the best way to protect one particular one from deletion, but yes you can rename them, ie see https://docs.oracle.com/cd/E19120-01/open.solaris/817-2271/gbion/index.html

A quick search and seems like its just a shell script: https://github.com/zfsonlinux/zfs-auto-snapshot
So you should be able to modify to your liking.

You might also be interested in: https://github.com/openzfs/zfs/issues/10946

**Comment by mailinglists35** [comment]
@jittygitty I don't think you understand the request. I want to be able to rename a snapshot by mv-ing the directory where it is mounted.

**Comment by jittygitty** [comment]
@mailinglists35 Sorry, you said: "...the mv command until I realized its not going to work and I would have to go to 'zfs rename' command." So seemed you found out its not possible and your last question I thought was asking for a way to "skip" deleting particular snapshots. I can confirm you can't rename snapshots with "mv", since then you could technically move it too. But like I said you, can still prevent auto deletion of some (by looking at zfs-auto-snapshot code etc), which seemed like your end goal.

**Comment by Dragon-debian** [comment]
As far as I know, zfs snapshots are available in `.zfs/snapshot` not in `.zfs/share`.
You can rename, create and destroy zfs snapshots by renaming, creating and destroying directories in `.zfs/snapshot`.
For that you have to set `zfs_admin_snapshot` parameter of zfs module to a nonzero integer. I have tested different values but it seems that every nonzero value has the same result. see `modinfo -F parm | grep zfs_admin_snapshot`.

Its possible to change modules parameters during run time using sysfs interface:
```
echo 1 > /sys/module/zfs/parameters/zfs_admin_snapshot
```
You can set it permanently by configuring it in `modprobe` configs.

After changing this, you can rename, create and remove snapshots in `.zfs/snapshot` with `mv`,`mkdir` and `rmdir` commands.
I tested those and `zfs list` showed that it really works. Interesting part is that `zpool history` didnt show anything about those changes.
If you run `ls` in `.zfs/snapshot`,  zfs mounts all snapshots on their respective directory, so then you cant destroy zfs snapshots with `rmdir`, the solution is to unmount the snapshot, then you can use `rmdir` to destroy it. this also applies to `mv`, if you want to rename a snapshot with `mv`, snapshots needs to be unmounted.

I didnt test that `zfs clone`, `zfs rollback`, `zfs send` and other zfs subcommands work or not but I think they do.
Im using debian 12 (bookworm) with `zfs V2.1.4`.
 I dont know how stable this feature is, use it at your own risk.

**Comment by GregorKopka** [comment]
> **Interesting part is that `zpool history` didnt show anything about those changes.**

This looks like a bug to me.

**Comment by mailinglists35** [comment]
>For that you have to set zfs_admin_snapshot parameter of zfs module to a nonzero integer

whoa awesome! I'll try it
